### PR TITLE
etcd: fix json unmarshall error and sources not found after upgrade to 2.0.2

### LINKED ISSUE #1644: Source unbound after upgrade
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.
cluster: 1 master, 2 workers, 2 sources
upgrade from v2.0.x to nightly
2. What did you expect to see?
upgrade success
3. What did you see instead?
list-member --worker
```
"worker": {
                "msg": "",
                "workers": [
                    {
                        "name": "dm-172.28.0.111-8262",
                        "addr": "172.28.0.111:8262",
                        "stage": "bound",
                        "source": "mysql-replica-02"
                    },
                    {
                        "name": "dm-172.28.0.112-8262",
                        "addr": "172.28.0.112:8262",
                        "stage": "free",
                        "source": ""
                    }
                ]
            }
```
4. Versions of the cluster

    - DM version (run `dmctl -V` or `dm-worker -V` or `dm-master -V`):

        ```
       nightly
        ```

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve? 

part of https://github.com/pingcap/dm/issues/1636

Info struct changed in v2.0.2, but forgot to update etcd value.

close #1644 
upgrade etcd before scheduler start

### What is changed and how it works?
Try unmarshall and convert old info to new info if unmarshall error happened.
Try upgrade etcd key before scheduler start.

### Check List 

Tests 

 - Unit test
 - upgrade test

### CODE REVIEW

**Comment by lance6716** [comment]
another bug is after DM recovers source bounds, it created grpc connections to every worker. but when scheduler closes due to the bug of this pr,  it doesn't close connections. we might fix this in another pr

**Review by Ehco1996 — COMMENTED** [review]
do we have a plan to clean  those forward compatible  code  like `oldinfo` ?

**Comment by GMHDBJD** [comment]
> do we have a plan to clean those forward compatible code like `oldinfo` ?

Maybe we can only remove them until we did not support the previous version, otherwise, we should make sure that upgrade via an old version successfully.

**Comment by lance6716** [comment]
(I'll create an issue to attach the "affected-v2.0.2" label, because we give the link to issue page not PR page in release note)

File: tests/tiup/upgrade-from-v2.sh [hunk 1]
```diff
@@ -46,6 +47,7 @@ function migrate_in_v2 {
     exec_incremental_stage2
 
     check_sync_diff $WORK_DIR $CUR/conf/diff_config.toml
+    check_sync_diff $WORK_DIR $CUR/conf/diff_config_optimistic.toml
```
Comment (lichunzhu) [inline tests/tiup/upgrade-from-v2.sh:50]: Can we check whether the lock is resolved here?

**Review by lichunzhu — COMMENTED** [review]
Rest LGTM

**Review by lance6716 — COMMENTED** [review]
we might add v2.0.2 to           
```
previous_v2: ["v2.0.0", "v2.0.1"]
```

File: pkg/shardddl/optimism/info.go [hunk 2]
```diff
@@ -148,6 +148,17 @@ func (i Info) toJSON() (string, error) {
...
 func infoFromJSON(s string) (i Info, err error) {
 	err = json.Unmarshal([]byte(s), &i)
+	if err != nil {
+		// For compatibility.
+		// In v2.0.2, we changed struct of table-info-after but forgot to upgrade etcd value.
+		// To keep the ModRevision of info, we change them after getting info instead of change all the value in etcd when upgrade
+		// All the Info will be upgraded after new info putted or lock resolved.
+		oldInfo, newErr := oldInfoFromJSON(s)
+		if newErr != nil {
+			return
```
Comment (lance6716) [inline pkg/shardddl/optimism/info.go:158]: we might log `newErr` here

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: pkg/upgrade/upgrade.go [hunk 3]
```diff
@@ -222,8 +270,6 @@ func upgradeToVer3(cli *clientv3.Client, uctx Context) error {
 			// note that we lost CreateRevision, Lease, ModRevision, Version
 			ops = append(ops, clientv3.OpPut(newKey, string(kv.Value)))
 		}
-		// delete old key to provide idempotence
-		ops = append(ops, clientv3.OpDelete(pair.old.Path(), clientv3.WithPrefix()))
```
Comment (lichunzhu) [inline pkg/upgrade/upgrade.go:226]: Why do we delete these?

Comment (GMHDBJD) [reply pkg/upgrade/upgrade.go:226]: for rollback to old version

Comment (GMHDBJD) [reply pkg/upgrade/upgrade.go:226]: Seems may cause wrong bound relation for later upgrade

Comment (GMHDBJD) [reply pkg/upgrade/upgrade.go]: revert

**Review by lichunzhu — COMMENTED** [review]
/lgtm

**Pull request merged**